File: dart/algorithms/sort/sort.dart
import 'package:collection/collection.dart';

import 'bubble_sort.dart';
import 'insertion_sort.dart';
import 'merging_sort.dart';

get input => [0, 20, 0, -1, 3, double.maxFinite.toInt(), 55];
get expected => [-1, 0, 0, 3, 20, 55, double.maxFinite.toInt()];

void main() {
  sortWithStopwatch(
    name: 'Bubble sort',
    input: input,
    expected: expected,
    operation: (getInput) => bubbleSort(input),
  );

  sortWithStopwatch(
    name: 'Insertion sort',
    input: input,
    expected: expected,
    operation: (input) => insertSort(input),
  );

  sortWithStopwatch(
    name: 'Merge sort',
    input: input,
    expected: expected,
    operation: (input) => mergingSort(input),
  );
}

void sortWithStopwatch({
  required String name,
  required List<int> input,
  required List<int> expected,
  required List<int> Function(List<int> input) operation,
}) {
  print(name);
  print('Unsorted list: $input');
  Stopwatch stopwatch = Stopwatch()..start();
  final result = operation(input);
  stopwatch.stop();
  print('Sorted list: $result');
  print('In time: ${stopwatch.elapsedMilliseconds} millis');
  assert(result.equals(expected));
  print('');
}

File: kotlin/algorithms/sort/Sort.kt
package algorithms.sort

import kotlin.system.measureTimeMillis

val input: List<Int> get() = listOf(0, 20, 0, -1, 3, Int.MAX_VALUE, 55)
val expected: List<Int> get() = listOf(-1, 0, 0, 3, 20, 55, Int.MAX_VALUE)

fun main() {
    sortWithTimeMillis(
        name = "Bubble sort",
        input = input,
        expected = expected,
    ) { bubbleSort(it) }

    sortWithTimeMillis(
        name = "Insertion sort",
        input = input,
        expected = expected,
    ) { insertionSort(it.toMutableList()) }

    sortWithTimeMillis(
        name = "Merge sort",
        input = input,
        expected = expected,
    ) { mergeSort(it.toMutableList()) }
}

private fun sortWithTimeMillis(
    name: String,
    input: List<Int>,
    expected: List<Int>,
    operation: (List<Int>) -> List<Int>,
) {
    println(name)
    println("Unsorted list: $input")
    var result: List<Int>
    val time = measureTimeMillis {
        result = operation(input)
    }
    println("Sorted list: $result")
    println("In time: $time millis")
    assert(result == expected)
    println()
}
